src/main/java/org/littleshoot/proxy/RateLimiter.java

```diff
@@ -0,0 +1,36 @@
+package org.littleshoot.proxy;
+
+import io.netty.handler.codec.http.HttpRequest;
+
+/**
+ * Interface for rate limiting requests
+ */
+public interface RateLimiter {
+
+  /**
+   * Rate Limiting general http requests
+   *
+   * @param httpRequest
+   * @return <code>true</code> if http request hit rate limit
+   * <code>false</code>.
+   */
+  boolean isOverLimit(HttpRequest httpRequest);
+
+  /**
+   * Rate Limiting user authentication requests
+   *
+   * @param username
+   * @return <code>true</code> if user hit rate limit
+   * <code>false</code>.
+   */
+  boolean isAuthenticationOverLimit(String username);
```

Comment:
change this implementation to take the http request instead. if we want to inspect other attributes on the request we should be able to do so. e.g. client ip or headers.

Reply:
I did thought this through and there is method like this above in the same interface:

```
public interface RateLimiter {
  boolean isOverLimit(HttpRequest httpRequest);    // <<<<<<<<<<<<<<<<<<<<<<<<
  boolean isAuthenticationOverLimit(String username);
  boolean isAuthenticationFailureOverLimit(String username);
}
```

^ first method. Because of our specific case with Lyft's rate limiter we have to call it twice that is why I've introduced 2 additional methods `isAuthenticationOverLimit` and `isAuthenticationFailureOverLimit`. Can't use one here.